discovery.md

```diff
@@ -0,0 +1,378 @@
+
+There are use cases where the Discovery API is served as a collection of other
+producers. These take two forms:
+
+1. Middleware
+2. Aggregator (or *Events Broker*)
+
+In the Middleware case, discovery is a combined set of producers for where the
+discovery is passed through largely unmodified and subscriber URIs are generally
+absolute (back to the producer).
+
+In the aggregator use case, the aggregator is expected to modify the output of
+discovery and subscription URIs will be relative to the aggregator itself.
+
+## Notations and Terminology
+
+### Notational Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+### Terminology
+
+This specification defines the following terms:
+
+### Producer
+
+The "producer" is a specific instance, process or device that creates the data
+structure describing the CloudEvent.
+
+### Intermediary
+
+An "intermediary" receives a message containing an event for the purpose of
+forwarding it to the next receiver, which might be another intermediary or a
+consumer. A typical task for an intermediary is to route the event to receivers
+based on the information in the event context.
+
+### Middleware
+
+Specific to discovery and subscription capabilities. Events middleware systems
+provide discovery services that may be directly passing through discovery from
+producers, or may aggregate or otherwise alter the discovery. This includes
+events brokers that may also be relaying discovery AND subscription information
+further upstream to event producers. Multiple middleware systems may exist
+between the producer and the consumer.
+
+### Consumer
+
+The entity receiving the event - the destination for the events related to a
+Subscription when the events are “pushed” by the Event Producer. In some
+instances this might be the same entity as the Subscriber. It could be the
+eventual event consumer or a pub/sub messaging entity which delivers the event
+to the eventual event consumer.
+
+### Subscription
+
+The request for events from an Event Producer or middleware system.
+
+### Event Subscriber
+
+The entity managing the lifecycle of a Subscription on behalf of an Event
+Consumer. In some instances this might be the same entity as the Event Consumer.
+
+## API Specification
+
+The discovery API takes the declarative API design approach. This document
+structure focuses on discovery where each document has a two part key of
+`producer` and `type`. This takes a denormalized view where the composite key
+of documents is the `producer` and `type`. Producer is defined as an arbitrary
+human readable string and type is the CloudEvent type attribute.
+
+// TODO
+// 1) need to rationalize fanout, broker aggregator scenario
+// 2) Does discovery need to be 2 levels? - producers -> types -> sources
```

Comment:
I wonder if it wouldn't be `producers -> source types -> types`

Reply:
I like `source types`. I've been thinking a lot about this event catalog topic over the past months, and I often get to the question, how to categorize and identify (groups of) sources.

Reply:
removed this todo, wasn't meant to be left in